[WIP] Provide the whole aggregation chain to aggregator server

mbohlool opened the pull request:
fixes #50011 

I like the approach here though I contemplated about passing the whole chain and assume the first in the chain is `kubeApiServer`/`delegateTarget` vs. passing `delegateTarget` separately. Tests may not pass yet as I prefer us to agree on the solution before doing cleanups/tests on this. 

Also the first commit is from the other PR and does not need review (at least not here :)

mbohlool committed:
Consolidate local OpenAPI specs and APIServices' spec into one data structure

mbohlool committed:
Provide the whole aggregation chain to aggregator server

sttts commented on cmd/kube-apiserver/app/server.go:
```diff
@@ -172,7 +172,9 @@ func CreateServerChain(runOptions *options.ServerRunOptions, stopCh <-chan struc
 	aggregatorConfig.ServiceResolver = serviceResolver
-	aggregatorServer, err := createAggregatorServer(aggregatorConfig, kubeAPIServer.GenericAPIServer, apiExtensionsServer.Informers)
+	aggregatorServer, err := createAggregatorServer(aggregatorConfig,
+		[]genericapiserver.DelegationTarget{kubeAPIServer.GenericAPIServer, apiExtensionsServer.GenericAPIServer},
```
Why not just make the delegate of an DelegationTarget available with `Delegate() DelegationTarget`? I don't like to have the chain here another time, after having it implicitly in the construction of each apiserver. I think the `Delegate()` method would give you exactly the same.

sttts replied:
And you wouldn't need those `len(...) < 1` check or to access `[0]`.

mbohlool replied:
good idea, will do.

sttts commented:
/cc @frobware

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -36,8 +36,9 @@ import (
-	specDownloadTimeout = 60 * time.Second
+	aggregatorUser                = "system:aggregator"
+	specDownloadTimeout           = 60 * time.Second
+	LocalDelegateChainNamePattern = "k8s_internal_local_delegation_chain_%d"
```
why public?

mbohlool replied:
will fix

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -47,36 +48,65 @@ type openAPIAggregator struct {
+	if err != nil {
+		return err
+	}
+	if filterPaths == nil {
```
why this default? nil should be no filter. Everything else is unexpected magic.

sttts replied:
do we actually need this method? It has side-effects and I wonder whether we should concentrate that to the logic below with the actual loop.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -110,6 +140,10 @@ type byPriority struct {
 func (a byPriority) Len() int      { return len(a.specs) }
 func (a byPriority) Swap(i, j int) { a.specs[i], a.specs[j] = a.specs[j], a.specs[i] }
 func (a byPriority) Less(i, j int) bool {
+	// All local specs will come first
```
if we cannot fix this in a better way, this deserves a huge warning that we deviate here from the API aggregation.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -47,36 +48,65 @@ type openAPIAggregator struct {
-		return nil, err
+	for i, chain := range delegateHandlers {
+		pathFilter := []string{"/apis/"}
+		// do not filter paths for the first spec. Use it as a base for merging.
```
add a "why" to the comment. Why not for the first spec?

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -47,36 +48,65 @@ type openAPIAggregator struct {
+	aggregator.FilterSpecByPaths(aggregatorOpenAPISpec, []string{"/apis/"})
+
+	// Reserving non-name spec for aggregator's Spec.
+	s.addLocalSpec(aggregatorOpenAPISpec, "")
```
technically the aggregator is in-front of the other servers. Shouldn't we start with its spec and add the others afterwards?

sttts replied:
ok, got it. "add" means adding to a map. So the order is not important here.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -132,6 +166,9 @@ func sortByPriority(specs []openAPISpecInfo) {
 		groupPriorities: map[string]int32{},
 	}
 	for _, spec := range specs {
+		if spec.apiService.Spec.Service == nil {
```
we skip local services in the order, but still add this special case in the comparison function?

mbohlool replied:
We don't skip them, we skip them only in computation of `groupPriorities`.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -47,36 +48,65 @@ type openAPIAggregator struct {
 
-	contextMapper request.RequestContextMapper
+func (s *openAPIAggregator) addLocalSpec(spec *spec.Swagger, name string) {
+	localApiService := apiregistration.APIService{}
```
why not just inline  `localApiService` in `openAPISpecInfo{ ... }`?

mbohlool replied:
It is just easier instead of creating `ObjectMeta`, `APIServiceSpec`, etc. just to change two field in them.

mbohlool replied:
Also I need to import ObjectMeta's package for example that is not necessary if we don't inline it.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -47,36 +48,65 @@ type openAPIAggregator struct {
+func (s *openAPIAggregator) addLocalSpec(spec *spec.Swagger, name string) {
+	localApiService := apiregistration.APIService{}
+	localApiService.Name = name
+	localApiService.Spec.Service = nil
```
nil is the zero value, no need to set it

mbohlool replied:
fixed.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -47,36 +48,65 @@ type openAPIAggregator struct {
+		if i == 0 {
+			pathFilter = []string{}
+		}
+		s.LoadAndAddLocalSpec(chain, fmt.Sprintf(LocalDelegateChainNamePattern, i), pathFilter)
```
error is ignored?

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -47,36 +48,65 @@ type openAPIAggregator struct {
 		return nil, err
 	}
 	// Remove any non-API endpoints from aggregator's spec. aggregatorOpenAPISpec
 	// is the source of truth for all non-api endpoints.
```
Cannot follow this comment. The filter call remove everything other than `/apis`. But then `aggregatorOpenAPISpec is the source of truth for all non-api endpoints`? But there is none left in `aggregatorOpenAPISpec`.

mbohlool replied:
fixed.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -141,15 +178,20 @@ func sortByPriority(specs []openAPISpecInfo) {
 func (s *openAPIAggregator) buildOpenAPISpec() (specToReturn *spec.Swagger, err error) {
-	specToReturn, err = aggregator.CloneSpec(s.inProcessDelegatesOpenAPISpec)
+	firstDelegateName := fmt.Sprintf(LocalDelegateChainNamePattern, 0)
+	localDelegateSpec, exists := s.openAPISpecs[firstDelegateName]
```
Couldn't we just start with an empty spec, sort by priority (also the first one) and then merge?

sttts replied:
Then also the special if-clause in the loop would go away and the first is not special anymore.

mbohlool replied:
The `MergeSpec` method is cloning `Source` spec. If we start with an empty one, we would clone first delegate spec as it is the largest one. I chose not to do it for performance.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -47,36 +48,65 @@ type openAPIAggregator struct {
-	contextMapper request.RequestContextMapper
+func (s *openAPIAggregator) addLocalSpec(spec *spec.Swagger, name string) {
+	localApiService := apiregistration.APIService{}
+	localApiService.Name = name
```
so this name is not actually the one from the auto registration controller of the aggregator, but some artifical one per delegation apiserver in the chain?

mbohlool replied:
Yes.

sttts commented:
A general question about https://github.com/kubernetes/kubernetes/pull/50896/files#diff-b0af072d7b063d5485d67d9dcbaf862eR295: does this mean we download the spec per APIService? So if a server provides 10 api groups with 2 versions each, we download, unmarshal, filter 20 times?

mbohlool commented:
To answer your last question, yes. we do that. We passed a handler and has no easy way to know if they are the same server from different APIServices. We can definitely improve that later.

I am closing this PR and add its code to #50864 as there are share codes between them and also the rebase would be hell if we move them forward separately. I addressed your comments here. Please continue discussion here if it make sense or move them to the other PR.
